internal/venice-common/src/main/java/com/linkedin/venice/memory/HeapSizeEstimator.java

```diff
@@ -0,0 +1,188 @@
+     */
+    ARRAY_HEADER_SIZE = roundUpToNearestAlignment(OBJECT_HEADER_SIZE + Integer.BYTES);
+    POINTER_SIZE = isCompressedOopsEnabled ? 4 : 8;
+  }
+
+  /**
+   * This function is a helper to make it easier to implement {@link Measurable#getSize()}. It provides the "overhead"
+   * of each instance of the given class, while making the following assumptions:
+   *
+   * 1. No pointer is null
+   * 2. Any variable sized objects (e.g., arrays and other collections) are empty
+   * 3. Polymorphism is ignored (i.e., we do not guess the size of any potential implementation of an abstraction)
+   *
+   * The intended usage of this function is to call it once per runtime per class of interest and to store the result in
+   * a static field. If the instances of this class (and any other class contained within it) contain no fields which
+   * are variable-sized object, no nullable objects, and no polymorphism, then the result of this function is
+   * effectively the heap size of each instance. If any of these conditions are not true, then a function should be
+   * implemented which uses this class overhead as a base and then makes adjustments, e.g.:
+   *
+   * 1. For any null pointer, we can subtract from the base a value equal to the output of this function for the type of
+   *    the field which is null.
+   * 2. For any variable-sized object, then the actual size has to be taken into account.
+   * 3. If polymorphism comes into play, then the specific sizes of actual instances need to take into account the
+   *    concrete classes they are made of.
+   *
+   * @throws {@link StackOverflowError} It should be noted that this function is recursive in nature and so any class
+   *         which contains a loop in its class graph will cause a stack overflow.
+   */
+  public static <T> int getClassOverhead(Class<T> c) {
+    if (c == null) {
+      throw new NullPointerException("The class param must not be null.");
+    }
+
+    if (c.isEnum()) {
+      /**
+       * Each enum value is stored statically, and so the per-instance overhead of enum fields is only the pointer size,
+       * already taken into account in the field loop below.
+       */
+      return 0;
+    }
+
+    if (c.isPrimitive()) {
+      return PRIMITIVE_SIZES.get(c);
+    }
+
+    int size = c.isArray() ? ARRAY_HEADER_SIZE : OBJECT_HEADER_SIZE;
+
+    size += overheadOfFields(c);
+
+    // TODO: Fix imprecision due to internal loss in parent class layouts
+    Class parentClass = c.getSuperclass();
+    while (parentClass != null) {
+      size += overheadOfFields(parentClass);
+      parentClass = parentClass.getSuperclass();
+    }
+
+    int finalSize = roundUpToNearestAlignment(size);
+
+    return finalSize;
+  }
+
+  /** Deal with alignment by rounding up to the nearest alignment boundary. */
+  private static int roundUpToNearestAlignment(int size) {
+    int partialAlignmentWindowUsage = size % ALIGNMENT_SIZE;
```

Comment:
Alignment could be modified with command line.

Reply:
I know that it can be configured at start time. But are you saying it can be modified at run time?

Reply:
No, it can't be changed at runtime, but you defined static alignment which might not be accurate. That's an edge case of course, I was just saying.

Reply:
Oh I see... you mean that alignment can be overridden by a JVM option, in which case it wouldn't decide just based on whether it's a 32 bits or 64 bits VM?

Is it via `-XX:ObjectAlignmentInBytes`?

Reply:
Yes, it is set statically at bootstrap, but code does not take this into account. Feel free to ignore that, since this is very rare edge case.